test/library/standard/FileSystem/chmod/changeMode.execopts

```diff
@@ -0,0 +1,38 @@
+--permissions 0o0600 # changeMode.rwu.good
+--permissions 384 # changeMode.rwu.good
+--permissions 0o0400 # changeMode.ru.good
+--permissions 256 # changeMode.ru.good
+--permissions 0o0200 # changeMode.wu.good
+--permissions 128 # changeMode.wu.good
+--permissions 0o0100 # changeMode.xu.good
+--permissions 64 # changeMode.xu.good
+--permissions 0o0700 # changeMode.rwxu.good
+--permissions 448 # changeMode.rwxu.good
+--permissions 0o0040 # changeMode.rg.good
+--permissions 32 # changeMode.rg.good
+--permissions 0o0020 # changeMode.wg.good
+--permissions 16 # changeMode.wg.good
+--permissions 0o0010 # changeMode.xg.good
+--permissions 8 # changeMode.xg.good
+--permissions 0o0070 # changeMode.rwxg.good
+--permissions 56 # changeMode.rwxg.good
+--permissions 0o0004 # changeMode.ro.good
+--permissions 4 # changeMode.ro.good
+--permissions 0o0002 # changeMode.wo.good
+--permissions 2 # changeMode.wo.good
+--permissions 0o0001 # changeMode.xo.good
+--permissions 1 # changeMode.xo.good
+--permissions 0o0007 # changeMode.rwxo.good
+--permissions 7 # changeMode.rwxo.good
+--permissions 0o4000 # changeMode.su.good
+--permissions 2048 # changeMode.su.good
+--permissions 0o2000 # changeMode.sg.good
+--permissions 1024 # changeMode.sg.good
+--permissions 0o1000 # changeMode.to.good
+--permissions 512 # changeMode.to.good
+--permissions 0o4700 # changeMode.rwsu.good
+--permissions 2496 # changeMode.rwsu.good
+--permissions 0o2070 # changeMode.rwsg.good
+--permissions 1080 # changeMode.rwsg.good
+--permissions 0o1007 # changeMode.rwto.good
+--permissions 519 # changeMode.rwto.good
```

Comment:
The above two lines fail on my system.  Instead of:
```
1007
1007
```
I get:
```
2070
2070
```

Reply:
Same for this one.

Reply:
I am assuming the last two lines for both the ones you mentioned did not run for some reason. The `prediff` finds the permissions of the file and thus it got back the permissions of the previous run (which matches the output you got on your machine). The not running might have something to do with the sticky bit, since it happened for the sticky bit test for both cases, but I cannot understand why it is happening.

Reply:
> did not run for some reason

It would have indicated that it did not run in some way.  Here's my output from `start_test`

```
... # earlier runs
[Executing system-wide preexec]
[Executing preexec changeMode.preexec]
[Executing program ./changeMode --permissions 1080 < /dev/null]
[Elapsed execution time for "library/standard/FileSystem/chmod/changeMode (execopts: 36)" - 0.017 seconds]
[Executing prediff ./changeMode.prediff]
[Executing diff changeMode.rwsg.good changeMode.0-36.exec.out.tmp]
[Success matching program output for library/standard/FileSystem/chmod/changeMode]
[Executing system-wide preexec]
[Executing preexec changeMode.preexec]
[Executing program ./changeMode --permissions 0o1007 < /dev/null]
[Elapsed execution time for "library/standard/FileSystem/chmod/changeMode (execopts: 37)" - 0.017 seconds]
[Executing prediff ./changeMode.prediff]
[Executing diff changeMode.rwto.good changeMode.0-37.exec.out.tmp]
1,2c1,2
< 1007
< 1007
---
> 2070
> 2070
[Error matching program output for library/standard/FileSystem/chmod/changeMode (execopts: 37)]
[Executing system-wide preexec]
[Executing preexec changeMode.preexec]
[Executing program ./changeMode --permissions 519 < /dev/null]
[Elapsed execution time for "library/standard/FileSystem/chmod/changeMode (execopts: 38)" - 0.016 seconds]
[Executing prediff ./changeMode.prediff]
[Executing diff changeMode.rwto.good changeMode.0-38.exec.out.tmp]
1,2c1,2
< 1007
< 1007
---
> 2070
> 2070
[Error matching program output for library/standard/FileSystem/chmod/changeMode (execopts: 38)]
```

Reply:
I manually compiled the program and ran it with `--permissions=xyz` and this was the output:

```
uncaught IllegalArgumentError: bad cast from string 'xyz' to int(64)
    <command line setting of 'permissions'>:0: thrown here
    <command line setting of 'permissions'>:0: uncaught here
```

I believe this is another test worth writing!

So then I put this line at the end of the `changeMode.execopts`:

`--permissions xyz # changeMode.rwto.good`

Then I ran the tests using `start_test` and it ran all the tests with success:

```
[Executing preexec changeMode.preexec]
[Executing program ./changeMode --permissions 519 < /dev/null]
[Elapsed execution time for "library/standard/FileSystem/chmod/changeMode (execopts: 38)" - 0.023 seconds]
[Executing prediff ./changeMode.prediff]
[Executing diff changeMode.rwto.good changeMode.0-38.exec.out.tmp]
[Success matching program output for library/standard/FileSystem/chmod/changeMode]
[Executing preexec changeMode.preexec]
[Executing program ./changeMode --permissions xyz < /dev/null]
[Elapsed execution time for "library/standard/FileSystem/chmod/changeMode (execopts: 39)" - 0.029 seconds]
[Executing prediff ./changeMode.prediff]
[Executing diff changeMode.rwto.good changeMode.0-39.exec.out.tmp]
[Success matching program output for library/standard/FileSystem/chmod/changeMode]
```

This suggests that even if it did not run correctly, it would not indicate it. I believe it ignores any errors that might be thrown during the file's execution.

Reply:
Ohhhh, that's interesting.  Well spotted!  I think that's to do with the .prediff file, then.  I want to say that we're overwriting the original output when running

Reply:
> I believe this is another test worth writing!

I agree :)

Reply:
At the end of running `start_test` with the given `changeMode.execopts`, the permissions on `file` and `dir` are:
```
-------rwt 1 <username> <group> 0 Mar 29 17:24 file
d------rwt 2 <username> <group> 4096 Mar 28 17:10 dir
```

If you manually compile `changeMode.chpl` and run it with `./<executable> --permissions=519`, does it give you an error? If not, does it set the permissions of `file` and `dir` correctly?

Reply:
I get:
```
uncaught SystemError: Invalid argument (in chmod with path "file")
  changeMode.chpl:7: thrown here
  changeMode.chpl:7: uncaught here
```

Reply:
That is very interesting. That is a type of error I have not gotten yet. Could you also tell me what happens if you run it with `--permissions=123456`? I do not get any error since, like you mentioned, it behaves like python's chmod function which takes the last 4 digits to set the permissions. The reason I want you to test this is because if I run `chmod 123456 <file>` directly on my shell, then I get an `chmod: invalid mode: '123456'` error which is similar to the error you have (invalid even though it is numbers).

Reply:
However, I still do not understand why this is happening. Could you possibly call in another person who could test this out and see if they get the same error?

Reply:
> Could you also tell me what happens if you run it with `--permissions=123456`?

I also get an error.

Interestingly, I also get an error with python:
```
python
Python 2.7.13 (default, Jan 11 2017, 10:56:06) [GCC] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.chmod("file", 448)
>>> os.chmod("file", 123456)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 22] Invalid argument: 'file'
```
And with Python 3
```
python3
Python 3.4.6 (default, Mar 01 2017, 16:52:22) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.chmod("file", 448)
>>> os.chmod("file", 123456)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 22] Invalid argument: 'file'
```

Reply:
As you can see, it works when a valid permission is given, but it complains about the file path when a bad permission is given (seems like I should report that to Python, they should improve their error messages)

Reply:
So python3 does not raise an error for me.

```
$ python3
Python 3.7.1 (default, Dec 14 2018, 19:28:38) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.chmod("file", 448)
>>> os.chmod("file", 123456)
>>> os.chmod("file", 519)
>>> exit()
$ ls -l file
-------rwt 1 <username> <group> 0 Apr  1 14:12 file
```

And neither does python2

```
$ python2
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.chmod("file", 448)
>>> os.chmod("file", 123456)
>>> os.chmod("file", 519)
>>> exit()
$ ls -l file
-------rwt 1 <username> <group> 0 Apr  1 14:12 file
```

The fact that it is recognizing invalid mode arguments is good (both for Python and Chapel), I'm unsure why it does not do that on my machine. But interestingly, both are not recognizing `519` or `0o1007` as a valid permission. That should not happen. Does it also raise an error if you run `chmod` on linux:
`chmod 1007 file`.
Does this raise an error? If yes, I would assume it is some problem with the sticky bit. If not, I am not sure again.

Reply:
```
$ chmod 1007 file
chmod: changing permissions of 'file': Invalid argument
```
I wonder if it's something a system administrator turns on or off?

Here's an interesting thing:
```
$ chmod 519 file
chmod: invalid mode: ‘519’
Try 'chmod --help' for more information.
```
That's a different error message (I'm assuming because 9 is meaningless to chmod)

Reply:
I did some research on the internet and it says that such an error usually occurs due to some sort of insufficient privilege. However, the surprising case is that it does not happen for other modes. Moreover, insufficient permissions is supposed to raise an `Operation not permitted` error. What would your advice be at this point? I believe we have narrowed down the problem but I am unsure of the course of action.

Reply:
Okay, here's my thinking: the test itself has shown that we are consistent with Python's behavior in all these different scenarios, which is great.  However, the output varies.  We know of two different cases and the behaviors that can occur.

So, I would split these cases from the main one and make a separate test to lock in their behavior.  This test would then check what the expected behavior for the system is (i.e. does Python allow it?  Then Chapel should allow it.  Does Python give an error?  Then Chapel should give an error).  It's going to be a more complicated test to write.  You might need to generate or choose the appropriate .good file based on what Python does using the .prediff helper, or you could use Chapel's Spawn library to create a Python subprocess and check if you get the same behavior in the Chapel program itself.

Reply:
If you're worried about that test case being a distraction from your proposal, you could make it happen in a separate PR (I would not hold it against you and it would allow us to merge this PR earlier, but I also won't hold this PR being still open against you either)